Sprint 12

maksimbelov95 opened the pull request:
Сдача спринта 12

maksimbelov95 committed:
Добавлен ImageListService
Реализована закачка и установка изображения в ячейки

maksimbelov95 committed:
Реализация структуры Photo

maksimbelov95 committed:
Добавлен сброс таски в ImageListService

maksimbelov95 committed:
Исправлена функция func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath), приводившая к ошибке "Out of range"

maksimbelov95 committed:
Добавлен выход из профиля с последующим переходом к окну авторизации

maksimbelov95 committed:
фикс лишней строчки кода

maksimbelov95 committed:
Добавлен DataFormatter.
Добавлен вывод даты из строчки createdAt в поле cell.dateLabel

maksimbelov95 committed:
Установлен likeButtonImage в соответствие с лайком пользователя

maksimbelov95 committed:
Исправлен action

maksimbelov95 committed:
Исправлен функционал замены фото с менящимся лайком
Убраны лишние принты

Nikitaaeee commented on ImageFeed/Auth/URLSession+Extension+NetworkError.swift:
```diff
@@ -51,11 +51,13 @@ extension URLSession{
             let session = URLSession.shared
             let task = session.dataTask(with: request, completionHandler: {data, response, error in
+                print(request)
                 if let data = data, let response = response, let statusCode = (response as?
```
**print'ы** могут быть очень полезны на момент написания кода или поиска багов, но в условный "продакшн" их лучше не пускать, можно удалять их перед отправкой работы на пулл-реквест 🙂

Nikitaaeee commented on ImageFeed/Auth/URLSession+Extension+NetworkError.swift:
```diff
@@ -51,11 +51,13 @@ extension URLSession{
                     if 200 ..< 300 ~= statusCode {
                         do {
                             let decoder = JSONDecoder()
+                            decoder.keyDecodingStrategy = .convertFromSnakeCase
```
👍🏻👍🏻👍🏻

Nikitaaeee commented on ImageFeed/ImageList/ImageListService.swift:
```diff
@@ -0,0 +1,138 @@
+        guard let request = makePhotosNextPage() else {
+            assertionFailure("Invalid request")
+            return
+        }
```
- Код выглядит немного "слипшимся"
Обычно принято оставлять одну пустую строку между методами/классами/структурами и другими блоками кода.
Также для организации кода можно добавлять комментарии типа `// MARK: - ...`
Подробнее про организацию кода можно почитать [в этой статье](https://habr.com/ru/articles/744470/), было бы здорово в дальнейшем оформлять код в таком стиле)

- Свойства лучше называть с маленькой буквы в **snakeCase**, так как с заглавной буквы в _Swift_ называют только классы, структуры, перечисления, протоколы.
В учебнике могло быть написано, что в случае `static let DidChangeNotification` можно использовать название с заглавной буквы, но в разработке чаще принято все таки называть такие свойства с маленькой.

Nikitaaeee commented on ImageFeed/ImageList/ImageListService.swift:
```diff
@@ -0,0 +1,138 @@
+                }
+            }
+            currentTask?.resume()
+        }else {return}
```
Сейчас проверка на `if currentTask == nil { ...... }` создает внутри себя большую вложенность, можно написать код немного по-другому, убрав эту вложенность из проверки, например так:
`if currentTask != nil { return } `
В таком случае, если currentTask будет существовать - компилятор сразу завершит выполнение метода.

Nikitaaeee commented on ImageFeed/ImageList/ImagesListCell.swift:
```diff
@@ -1,8 +1,26 @@
     @IBOutlet var likeButton: UIButton!
     @IBOutlet var dateLabel: UILabel!
+    
+    @IBAction func likeButtonTapped(_ sender: Any) {
+        delegate?.imageListCellDidTapLike(self)
+    }
```
Можно сделать этот метод приватным, так как он вызывается только в области видимости класса ячейки.

Nikitaaeee commented on ImageFeed/ImageList/ImagesListViewController.swift:
```diff
@@ -1,86 +1,141 @@
-    private let photosName: [String] = Array(0..<20).map{ "\($0)" }
+    private var photos: [Photo] = []
+    
+    private lazy var dateFormatter: DateFormatter = {
+            let formatter = DateFormatter()
+            formatter.dateFormat = "dd MMMM yyyy"
+            return formatter
+        }()
```
Тут немного поплыли отступы, в большинстве случаев автоматическое выравнивание помогает исправить отступы:
1. `Command + A` - выделение всего кода
2. `Control + I` - автоматическое выравнивание

maksimbelov95 committed:
1) Добавлен alert на вопрос выхода из профиле
2) создание fateForrmater изменена
3) попытка изменить цвет ProgressiveHUD в appDelefate. но безуспешная, буду рад подсказке)

maksimbelov95 committed:
no message
